### PR TITLE
Make cross-compiling with the crossSystem argument work (v2)

### PR DESCRIPTION
**Pull request by DavidEGrayson**
This is version 2 of my previous pull request (https://github.com/NixOS/nixpkgs/pull/18386).  Compared to my previous pull request, this has a bit of a smaller scope: it does not add a `test` directory and it does not try to fix the deprecated `--with-headers` option used to compile GCC cross-compilers.  Also, it does not do as much fixing of the top-level logic, thanks to the merged pull request https://github.com/NixOS/nixpkgs/pull/19940 by @Ericson2314 which cleaned up a lot of the problems in that area.

This pull request fixes several things about cross-compiling with the `crossSystem` argument:

- It fixes the expressions for several derivations (gcc5, boehmgc, libffi, libxml2, readline, libtool, libiconv) so that their native derivations do not change when the top-level crossSystem argument is provided.
- It avoids using bootstrap tools to compile the cross-compilers; it uses the normal stdenv.

To accomplish these goals, some new things were added:

- stdenv.mkDerivationWithCrossArg
- lib.overrideNativeDrv
- lib.overrideNativeDrvs

###### Motivation for this change

For motivation information, see my comment on the old pull request: https://github.com/NixOS/nixpkgs/pull/18386#issuecomment-245522662

###### Things done to test

This pull request was tested using the expression here:

https://gist.github.com/DavidEGrayson/8f962b3ec6edf06365ecb0c70a88029f

The tests passed, all the derivations were successfully built, and the rpiHello derivation was successfully run on a Raspberry Pi.  This pull request should not cause a mass rebuild of native derivations.

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

######  Future work

Here are some similar things that could be done in the future:

- Search for more improper uses of `stdenv.cross`, possibly remove `stdenv.cross` to prevent future errors.
- Stop using the deprecated --with-headers option to configure GCC cross-compilers.
- Apply these GCC fixes to the GCC 6 expression.
- Get cross-compiling working on Darwin too; this pull request is mainly about getting it to work on Linux.
- Add automated tests to make sure the problems fixed by this PR do not reoccur.

### CODE REVIEW

**Comment by DavidEGrayson** [comment]
CC @shlevy and @bjornfor because they both asked for status updates on the previous pull request, and @dezgeg because he helped make the previous pull request work.  Also, @viric and @rasendubi.

**Comment by Ericson2314** [comment]
@DavidEGrayson you pinged me mentioning conflicts with my https://github.com/NixOS/nixpkgs/pull/21268 . Well besides the files changed there is a difference in approach. Currently in nixpkgs, as we both know, there is a only a very sloppy attempt to make nativeDrvs and crossDrvs. You clean that up with `overrideNativeDrv`, whereas I throw it out altogether. So were I to rebase mine on yours, I'd need to revert using those two lib functions (there's not bad, just no longer have any use), and changes like `pythonSupportReally`, whereas if you revert on mine, your diff should just shrink dramatically with no need to revert any of my changes.

**Comment by Ericson2314** [comment]
On a diffferent note, it would be amazing to get your tests into https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/release-cross.nix so they (I think) will be tested by hydra.

**Comment by DavidEGrayson** [comment]
I posted this pull request one day before Christmas Eve.  Now that the holidays are over, maybe a nixpkgs member would be interested in looking at it and merging it?  There was a lot of interest on my previous pull request (https://github.com/NixOS/nixpkgs/pull/18386) and I thought people were gung-ho to get it merged in.

A lot of this pull request is just fixing logic errors in the recipes for individual packages that are due to people misunderstanding how `stdenv.cross` works.

**Comment by Mic92** [comment]
I currently have a good understanding on how stdenv works.
However I want nixos on my ARM boards too, so I do my best and give it a shoot.
When rebasing your pull request against current master, 
I get the following evaluation error after, when trying to rebuild my system:

```
building Nix...
building the system configuration...
error: value is null while a Boolean was expected, at /etc/nixos/nixpkgs/pkgs/development/libraries/libxslt/default.nix:7:8
(use ‘--show-trace’ to show detailed location information)
```

This is the failing file: nixpkgs/pkgs/development/libraries/libxslt/default.nix

```
assert pythonSupport -> python2 != null;
assert pythonSupport -> libxml2.pythonSupport; # <--- this is failing line
```

you changed in libxml2/default.nix

```patch
-, pythonSupport ? (! stdenv ? cross) }:
+, pythonSupport ? null }:
```

As I said don't know enough about stdenv and cross compilation in this context to know what the right fix in this case would be...

**Comment by DavidEGrayson** [comment]
@Mic92 Thanks, that would have to be fixed before we merge this in.  If you're looking for a quick solution, maybe just remove the assertion that is failing.  For the longer term, I'd have to look into populating `libxml2.pythonSupport` with a proper boolean variable that says whether it supports Python or not.

**Comment by Mic92** [comment]
Unfortunately this is a mass-rebuild on NixOS. I currently do not have the resources locally to test this here. Probably someone with access to hydra could take a look at it.

**Comment by DavidEGrayson** [comment]
Hmmm, I have tests and code to try to make sure that this PR would not be a mass rebuild.  What is the first derivation that is getting rebuilt for you when you try it out?

**Review by Ericson2314 — CHANGES_REQUESTED** [review]
@DavidEGrayson Sorry for being slow. While I prefer the more general approach of my PR long term, that one is somewhat blocked on review. Also it would be nice to get cross into a better stage for my own comparing hashes in that one.

I don't want to penalize you for mine being blocked, so even though my PRs would mean undoing some of the work here, I'm now willing to merge this as soon as you

1. make these changes

2. make as many tests as possible part of `release-cross.nix` (recall my https://github.com/NixOS/nixpkgs/pull/21414). For example, or `rpiCrossSystem` is slightly different right? Maybe we want both variants in there? Maybe put the rest in `maintainers/scripts` (e.g. the HEAD, HEAD^ comparing ones)?

It would be great to get the green light from Travis too, but I understand that's often broken for silly reasons.

**Comment by Mic92** [comment]
@DavidEGrayson this would be the rebuild log, when doing `nixos-rebuild dry-build`:
https://gist.github.com/a92f9e47bf87f801034221aa48fa3647

**Comment by DavidEGrayson** [comment]
I am closing this.  @Ericson2314 is [working hard] to clean up cross-compiling in nixpkgs, so this pull request would probably just get in his way or make things harder.

Instead of working on this stuff, I am pursuing a project called [nixcrpkgs] where I will make my own Nix expressions for cross-compiling and just use nixpkgs as the base platform for my build system.  Since I don't need to worry about supporting a million different use cases, my expressions can be much cleaner and simpler than what has accumulated in nixpkgs over the years, especially when it comes to building a GCC cross-compiler. 

[working hard]: https://github.com/NixOS/nixpkgs/pulls?utf8=%E2%9C%93&q=author%3AEricson2314%20
[nixcrpkgs]: https://github.com/DavidEGrayson/nixcrpkgs

**Comment by Ericson2314** [comment]
@DavidEGrayson Thanks, definitely watching that repo. I had opened https://github.com/NixOS/nixpkgs/issues/21471 which I hope your repo + my nixpkgs PRs will greatly accelerate.

**Pull request closed without merging**
